rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,427 @@
+---
+authors: Tim Ross (tim.ross@goteleport.com)
+state: draft
+---
+
+# RFD 126 - Resource Migrations
+
+## Required Approvers
+
+- Engineering @zmb3 && (@fspmarshall || @espadolini )
+- Product: ( @klizhentas || @russjones )
+
+## What
+
+Providing support for running different versions of Auth simultaneously.
+
+## Why
+
+The upgrade process requires scaling Auth down to a single instance to ensure that migrations are only performed once as
+well as preventing a new version and an old version from operating on the same keys with different schema. This is
+cumbersome and makes the upgrade process a pain point for cluster admins. Scaling Auth down and back up also results in
+an [uneven load](https://github.com/gravitational/teleport/issues/7029) which can cause connectivity issues and backend
+latency that can result in a thundering herd.
+
+Migrations are applied during Auth initialization and prevent the instance from becoming ready and able to serve
+requests which causes downtime. To date, migrations have largely been applied in place on an existing key range, even
+for breaking changes, which prevents an older versions of Auth from operating on the data. This can result in data loss
+as new and older versions try to write different versions of a resource to the same key or worse, rendering older
+versions of Teleport unable to comprehend the data stored in the backend.
+
+## Details
+
+Instead of all migrations being a one-shot operation which break compatibility with previous versions of Auth we can
+leverage a phased approach which separates backend changes from application logic changes to allow Auth vN and Auth vN-1
+to coexist. This approach is, however, likely best suited for Cloud where we control the cadence and order of Teleport
+releases that are applied. Self-hosted customers should still follow the traditional upgrade procedure.
+
+To prevent having to split the changes across separate commits and ensure that they land in the correct release, we can
+leverage feature flags to allow all the changes to land at once and toggle the behavior when appropriate. Once the
+migration has been completed, the feature flags and old code can be removed. For self-hosted customers migrations should
```

Comment:
I think this could use some more details.  E.g. would these be "live" feature flags set by cluster administrators (e.g. env/cli options), or would these be built-in feature flags that would be toggled by followup PRs?